specs/dwritecore/DWriteCore.md

```diff
@@ -0,0 +1,255 @@
+{
+    DWRITE_BITMAP_DATA_BGRA32 result = {};
+
+    // Query the bitmap render target for the new interface.
+    ComPtr<IDWriteBitmapRenderTarget2> renderTarget2;
+    HRESULT hr = renderTarget->QueryInterface(&renderTarget2);
+    if (SUCCEEDED(hr))
+    {
+        // IDWriteBitmapRenderTarget2 exists, so we can get the bitmap the easy way.
+        hr = renderTarget2->GetBitmapData(OUT &result);
+        if (FAILED(hr))
+        {
+            throw Exception{ hr };
+        }
+    }
+    else
+    {
+        // We're using an older version that doesn't implement IDWriteBitmapRenderTarget2,
+        // so we have to get the bitmap by going through GDI. First get the bitmap handle.
+        HDC hdc = renderTarget_->GetMemoryDC();
+        HGDIOBJ dibHandle = GetCurrentObject(hdc, OBJ_BITMAP);
+        if (dibHandle == nullptr)
+        {
+            throw Exception{ HRESULT_FROM_WIN32(GetLastError()) };
+        }
+
+        // Call a GDI function to fill in the DIBSECTION structure for the bitmap.
+        DIBSECTION dib;
+        if (!GetObjectW(dibHandle, sizeof(dib), &dib))
+        {
+            throw Exception{ HRESULT_FROM_WIN32(GetLastError()) };
+        }
+
+        result.width = dib.dsBm.bmWidth;
+        result.height = dib.dsBm.bmHeight;
+        result.pixels = static_cast<uint32_t*>(dib.dsBm.bmBits);
+    }
+
+    return result;
+}
+```
+
+# API Details
+
+```c++
+/// <summary>
+/// Specifies the type of DirectWrite factory object.
+/// DirectWrite factory contains internal state such as font loader registration and cached font data.
+/// In most cases it is recommended to use the shared factory object, because it allows multiple components
+/// that use DirectWrite to share internal DirectWrite state and reduce memory usage.
+/// However, there are cases when it is desirable to reduce the impact of a component,
+/// such as a plug-in from an untrusted source, on the rest of the process by sandboxing and isolating it
+/// from the rest of the process components. In such cases, it is recommended to use an isolated factory 
+/// for the sandboxed component.
+/// </summary>
+enum DWRITE_FACTORY_TYPE
+{
+    /// <summary>
+    /// Shared factory allow for re-use of cached font data across multiple in process components.
+    /// Such factories also take advantage of cross process font caching components for better performance.
+    /// </summary>
+    DWRITE_FACTORY_TYPE_SHARED,
+
+    /// <summary>
+    /// Objects created from an isolated factory do not modify internal state or cached data used by
+    /// objects from other factories.
+    /// </summary>
+    DWRITE_FACTORY_TYPE_ISOLATED,
+
+#if DWRITE_CORE
+    /// <summary>
+    /// Objects created from a restricted factory do not use or modify internal state or cached data
+    /// used by other factories. In addition, the system font collection contains only well-known fonts.
+    /// </summary>
+    DWRITE_FACTORY_TYPE_RESTRICTED
```

Comment:
DISCUSSION: If we add something else to this list, a new factory type? Or do we have to "reunify" first?

CONSIDER: Figure out how to annotate that this is related to the Windows public header so they can align enum literal values.

Reply:
Regarding WIL or C++/WinRT, I think the raw HGDIOBJ is fine here because there's no resource ownership. An error-checking helper would be nice, but check_result takes an HRESULT, whereas what's needed in this case is a null check followed by GetLastError on failure.

Anyway, this code path in the sample snippet is meant to show how the "old way" (in the absence of the new API) requires jumping through some hoops, so maybe it's OK if the code is slightly more verbose. :-)

Reply:
I added a comment about keeping the enums consistent between future versions of this header in DWriteCore vs. the Windows SDK.